src/ser/raw/document_serializer.rs

```diff
@@ -0,0 +1,323 @@
+            root_serializer: &mut *self.root_serializer,
+        })?;
+
+        self.num_keys_serialized += 1;
+        Ok(())
+    }
+
+    pub(crate) fn end_doc(self) -> crate::ser::Result<DocumentSerializationResult<'a>> {
+        self.root_serializer.bytes.push(0);
+        let length = (self.root_serializer.bytes.len() - self.start) as i32;
+        self.root_serializer.replace_i32(self.start, length);
+        Ok(DocumentSerializationResult {
+            root_serializer: self.root_serializer,
+        })
+    }
+}
+
+impl<'a> serde::ser::SerializeSeq for DocumentSerializer<'a> {
+    type Ok = ();
+    type Error = Error;
+
+    fn serialize_element<T: ?Sized>(&mut self, value: &T) -> Result<()>
+    where
+        T: serde::Serialize,
+    {
+        self.serialize_doc_key(&self.num_keys_serialized.to_string())?;
+        value.serialize(&mut *self.root_serializer)
+    }
+
+    fn end(self) -> Result<Self::Ok> {
+        self.end_doc().map(|_| ())
+    }
+}
+
+impl<'a> serde::ser::SerializeMap for DocumentSerializer<'a> {
+    type Ok = ();
+
+    type Error = Error;
+
+    fn serialize_key<T: ?Sized>(&mut self, key: &T) -> Result<()>
+    where
+        T: serde::Serialize,
+    {
+        self.serialize_doc_key(key)
+    }
+
+    fn serialize_value<T: ?Sized>(&mut self, value: &T) -> Result<()>
+    where
+        T: serde::Serialize,
+    {
+        value.serialize(&mut *self.root_serializer)
+    }
+
+    fn end(self) -> Result<Self::Ok> {
+        self.end_doc().map(|_| ())
+    }
+}
+
+impl<'a> serde::ser::SerializeStruct for DocumentSerializer<'a> {
+    type Ok = ();
+
+    type Error = Error;
+
+    fn serialize_field<T: ?Sized>(&mut self, key: &'static str, value: &T) -> Result<()>
+    where
+        T: serde::Serialize,
+    {
+        self.serialize_doc_key(key)?;
+        value.serialize(&mut *self.root_serializer)
+    }
+
+    fn end(self) -> Result<Self::Ok> {
+        self.end_doc().map(|_| ())
+    }
+}
+
+impl<'a> serde::ser::SerializeTuple for DocumentSerializer<'a> {
+    type Ok = ();
+
+    type Error = Error;
+
+    fn serialize_element<T: ?Sized>(&mut self, value: &T) -> Result<()>
+    where
+        T: serde::Serialize,
+    {
+        self.serialize_doc_key(&self.num_keys_serialized.to_string())?;
+        value.serialize(&mut *self.root_serializer)
+    }
+
+    fn end(self) -> Result<Self::Ok> {
+        self.end_doc().map(|_| ())
+    }
+}
+
+impl<'a> serde::ser::SerializeTupleStruct for DocumentSerializer<'a> {
+    type Ok = ();
+
+    type Error = Error;
+
+    fn serialize_field<T: ?Sized>(&mut self, value: &T) -> Result<()>
+    where
+        T: serde::Serialize,
+    {
+        self.serialize_doc_key(&self.num_keys_serialized.to_string())?;
+        value.serialize(&mut *self.root_serializer)
+    }
+
+    fn end(self) -> Result<Self::Ok> {
+        self.end_doc().map(|_| ())
+    }
+}
+
+/// Serializer used specifically for serializing document keys.
+struct KeySerializer<'a> {
```

Comment:
There's a whole lot of boilerplate involved in implementing this type, since unfortunately there isn't a serialization equivalent to the `forward_to_deserialize_any!` macro in deserialization.

All but a few of these are basically unimplemented, the few being `serialize_str` and `serialize_some`.